target_decisioning_engine/tests/test_artifact_provider.py

```diff
@@ -0,0 +1,257 @@
+
+    def test_resume_polling(self):
+        with patch.object(ArtifactProvider, "fetch_artifact", return_value=None):
+            self.provider = ArtifactProvider(self.default_config)
+            self.provider.initialize()  # starts polling thread
+            self.provider.stop_all_polling()
+            self.assertTrue(self.provider.polling_halted)
+
+            self.provider.resume_polling()
+            self.assertTrue(self.provider.polling_timer.is_alive())
+            self.assertFalse(self.provider.polling_halted)
+
+    def test_schedule_next_update_polling_halted(self):
+        with patch.object(ArtifactProvider, "fetch_artifact", return_value=None):
+            self.provider = ArtifactProvider(self.default_config)
+            self.provider.initialize()  # starts polling thread
+            self.assertTrue(self.provider.polling_timer.is_alive())
+            self.assertFalse(self.provider.polling_halted)
+
+            self.provider.stop_all_polling()
+            self.provider.schedule_next_update()
+            self.assertIsNone(self.provider.polling_timer)
+            self.assertTrue(self.provider.polling_halted)
+
+    def test_fetch_artifact_with_etag(self):
+        artifact_url = "http://my.artifact.com"
+        request_mock = Mock()
+        self.provider = ArtifactProvider(self.default_config)
+        self.provider.last_response_etag = "123"
+        with patch("target_decisioning_engine.artifact_provider.http.request", request_mock):
+            self.provider.fetch_artifact(artifact_url)
+            expected_args = ("GET", "http://my.artifact.com")
+            expected_kwargs = {
+                "headers": {"If-None-Match": "123"},
+                "retries": HTTP_RETRY
+            }
+            request_mock.assert_called_once_with(*expected_args, **expected_kwargs)
+
+    def test_fetch_artifact_status_not_modified(self):
+        artifact_url = "http://my.artifact.com"
+        response_mock = Mock(status=NOT_MODIFIED)
+        request_mock = Mock(return_value=response_mock)
+        self.provider = ArtifactProvider(self.default_config)
+        self.provider.last_response_data = {"z": 99}
+        with patch("target_decisioning_engine.artifact_provider.http.request", request_mock):
+            self.provider.fetch_artifact(artifact_url)
+            self.assertEqual(self.provider.last_response_data, {"z": 99})
+
+    def test_fetch_artifact_status_ok_with_etag(self):
+        artifact_url = "http://my.artifact.com"
+        response_data = {"y": 88, "q": 14}
+        response_headers = {
+            "Etag": "12345"
+        }
+        response_mock = Mock(status=OK, data=json.dumps(response_data), headers=response_headers)
+        request_mock = Mock(return_value=response_mock)
+        self.provider = ArtifactProvider(self.default_config)
+        with patch("target_decisioning_engine.artifact_provider.http.request", request_mock):
+            artifact = self.provider.fetch_artifact(artifact_url)
```

Comment:
`fetch_artifact` should be private and not callable here.  If you look at the implementation of ArtifactProvider in JS, you will notice that all the implementation details for fetching the artifact are a black box to other objects.  You cannot tell the ArtifactProvider to "fetch an artifact".  Instead, you let ArtifactProvider manage all of the fetching logic as well as scheduling.  External classes simply subscribe to what has happened.